metadatastore/doc.py

```diff
@@ -0,0 +1,224 @@
+    return '{ago} ({date})'.format(ago=ago, date=dt)
+
+
+def _format_dict(value, name_width, value_width, name, tabs=0):
+    ret = ''
+    for k, v in six.iteritems(value):
+        if isinstance(v, collections.Mapping):
+            ret += _format_dict(v, name_width, value_width, k, tabs=tabs+1)
+        else:
+            ret += ("\n%s%-{}s: %-{}s".format(
+                name_width, value_width) % ('  '*tabs, k[:16], v))
+    return ret
+
+
+def _format_data_keys_dict(data_keys_dict):
+    from prettytable import PrettyTable
+
+    fields = reduce(set.union,
+                    (set(v) for v in six.itervalues(data_keys_dict)))
+    fields = sorted(list(fields))
+    table = PrettyTable(["data keys"] + list(fields))
+    table.align["data keys"] = 'l'
+    table.padding_width = 1
+    for data_key, key_dict in sorted(data_keys_dict.items()):
+        row = [data_key]
+        for fld in fields:
+            row.append(key_dict.get(fld, ''))
+        table.add_row(row)
+    return table
+
+
+def vstr(doc, indent=0):
+    """Recursive document walker and formatter
+
+    Parameters
+    ----------_
+    name : str, optional
+        Document header name. Defaults to ``doc._name``
+    indent : int, optional
+        The indentation level. Defaults to starting at 0 and adding one tab
+        per recursion level
+    """
+    headings = [
+        # characters recommended as headers by ReST docs
+        '=', '-', '`', ':', '.', "'", '"', '~', '^', '_', '*', '+', '#',
+        # all other valid header characters according to ReST docs
+        '!', '$', '%', '&', '(', ')', ',', '/', ';', '<', '>', '?', '@',
+        '[', '\\', ']', '{', '|', '}'
+    ]
+    name = doc['_name']
+
+    ret = "\n%s\n%s" % (name, headings[indent]*len(name))
+
+    documents = []
+    name_width = 16
+    value_width = 40
+    for name, value in sorted(doc.items()):
+        if name == 'descriptors':
+            # this case is to deal with Headers from databroker
+            for val in value:
+                documents.append((name, val))
+        elif name == 'data_keys':
+            ret += "\n%s" % str(_format_data_keys_dict(value))
+        elif isinstance(value, collections.Mapping):
+            if '_name' in value:
+                documents.append((name, value))
+            else:
+                # format dicts reasonably
+                ret += "\n%-{}s:".format(name_width, value_width) % (name)
+                ret += _format_dict(value, name_width, value_width,
+                                    name, tabs=1)
+        else:
+            ret += ("\n%-{}s: %-{}s".format(name_width, value_width) %
+                    (name[:16], value))
+    for name, value in documents:
+        ret += "\n%s" % (vstr(value, indent+1))
+        # ret += "\n"
+    ret = ret.split('\n')
+    ret = ["%s%s" % ('  '*indent, line) for line in ret]
+    ret = "\n".join(ret)
+    return ret
+
+
+def ref_doc_to_uid(doc, field):
```

Comment:
Is it at all sensible to just replace all the ref docs with their uids?  It would be simple enough to check all entries of the Document for other Documents